File: lab-4/src/abstracts/Alive.java
package abstracts;

import com.company.Action;
import com.company.Command;
import com.company.Decision;
import com.company.Quality;
import types.Feeling;

import java.util.Objects;

public abstract class Alive {
    protected String name;
    protected Decision decision;

    public Alive(String name) {
        this.name = name;
        this.decision = new Decision("");
    }

    public Action toLet(Alive entity) {
        System.out.println(name + " пропустил(а) " + entity.getName());
        return new Action(name + " пропустил(а) кого-то по имени " + entity.getName(),
                new Quality(
                        "Пропустил(а)"));
    }

    public Action toClimbinto(Thing thing) {
        return new Action(name + "полез в" + thing.getName(), new Quality("Полез"));
    }

    public void toFeel(Feeling feeling) {
        System.out.println(name + " почувствовал(а) " + feeling);
    }

    public void toFeel(Action action) {
        System.out.println(name + " почувствовал(а) как " + action.getDescription());
    }

    public void toDecide(String decision) {
        this.decision = new Decision(decision);
        System.out.println(name + " решил(а) " + decision);
    }

    public String toSee(Action action) {
        System.out.println(name + " увидел(а) как " + action.getDescription());
        return name + "увидел(а)" + action.getDescription();
    }

    public Action toShout(String cmd){
        System.out.println(name+" скомандавал(а) "+cmd);
        return new Action(name+" скомандавал(а) "+cmd, new Quality(
                "Команда "+cmd));
    }

    public Action toTake(Thing thing){
        System.out.println(name+" взял(а) "+thing.getName());
        return new Action(name+" взял(а) "+thing.getName(),
                new Quality("Взяли "+thing.getName()));
    }

    public Action toHear(Action action){
        System.out.println(name+" услышал(а), как "+action.getDescription());
        return new Action(name+" услышал(а), как "+action.getDescription(),
                new Quality("Услышали: "+action.getDescription()));
    }

    public String getName() {
        return name;
    }

    public Decision getDecision() {
        return decision;
    }

    @Override
    public String toString() {
        return name;
    }


    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Alive alive = (Alive) o;
        return name.equals(alive.name) && decision.equals(alive.decision);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, decision);
    }


}

File: lab-4-kotlin/src/abstracts/Alive.kt
package abstracts

import abstracts.Thing
import abstracts.Closable
import abstracts.Alive
import kotlin.Throws
import exc.IncorrectFileNameException
import kotlin.jvm.JvmStatic
import exc.NogivenNameException
import com.company.things.Rocket
import com.company.things.Shots
import com.company.things.Bullets
import com.company.things.Something
import com.company.things.ShirtSleev
import com.company.things.Bloodstain
import com.company.things.Door
import com.company.things.Healpack
import types.Feeling
import java.lang.ArrayIndexOutOfBoundsException
import java.io.File
import java.io.FileNotFoundException
import java.util.HashSet
import com.company.Person.Face
import java.util.Objects
import abstracts.Paintable
import com.company.*

abstract class Alive(var name: String) {
    var decision: Decision
        protected set

    init {
        decision = Decision("")
    }

    fun toLet(entity: Alive): Action {
        println(name + " пропустил(а) " + entity.name)
        return Action(
            name + " пропустил(а) кого-то по имени " + entity.name,
            Quality(
                "Пропустил(а)"
            )
        )
    }

    fun toClimbinto(thing: Thing): Action {
        return Action(name + "полез в" + thing.getName(), Quality("Полез"))
    }

    fun toFeel(feeling: Feeling) {
        println("$name почувствовал(а) $feeling")
    }

    fun toFeel(action: Action) {
        println(name + " почувствовал(а) как " + action.description)
    }

    fun toDecide(decision: String) {
        this.decision = Decision(decision)
        println("$name решил(а) $decision")
    }

    fun toSee(action: Action): String {
        println(name + " увидел(а) как " + action.description)
        return name + "увидел(а)" + action.description
    }

    fun toShout(cmd: String): Action {
        println("$name скомандавал(а) $cmd")
        return Action(
            "$name скомандавал(а) $cmd", Quality(
                "Команда $cmd"
            )
        )
    }

    fun toTake(thing: Thing): Action {
        println(name + " взял(а) " + thing.getName())
        return Action(
            name + " взял(а) " + thing.getName(),
            Quality("Взяли " + thing.getName())
        )
    }

    fun toHear(action: Action): Action {
        println(name + " услышал(а), как " + action.description)
        return Action(
            name + " услышал(а), как " + action.description,
            Quality("Услышали: " + action.description)
        )
    }

    override fun toString(): String {
        return name
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val alive = o as Alive
        return name == alive.name && decision == alive.decision
    }

    override fun hashCode(): Int {
        return Objects.hash(name, decision)
    }
}